File: apps/server-admin/src/lib/ddns/providers/duckdns.ts
import type { DDNSProviderContext, DDNSProviderDefinition, DDNSUpdateResult } from "../types";
import { ddnsProviderT, getTimeoutMs, parseTextResponse } from "./helpers";

const DUCKDNS_ENDPOINT = "https://ddns.duckdns.fnknock.cn/";
const duckdnsT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("duckdns", key, params);

export const duckdnsProvider: DDNSProviderDefinition = {
  name: "duckdns",
  label: "DuckDNS",
  fields: [
    {
      key: "domains",
      label: duckdnsT("fields.domains.label"),
      type: "text",
      placeholder: "home,lab",
      required: true,
      description: duckdnsT("fields.domains.description"),
    },
    {
      key: "token",
      label: "Token",
      type: "password",
      placeholder: "DuckDNS Token",
      required: true,
      description: duckdnsT("fields.token.description"),
    },
  ],
};

export async function duckdnsUpdate(
  { config, http }: DDNSProviderContext,
  ipv4: string | null,
  ipv6: string | null,
): Promise<DDNSUpdateResult> {
  const domains = config.domains?.trim();
  const token = config.token?.trim();

  if (!domains || !token) {
    return { success: false, message: duckdnsT("configIncomplete") };
  }

  if (!ipv4 && !ipv6) {
    return { success: false, message: duckdnsT("noIpAvailable") };
  }

  const payload = {
    domains,
    token,
    ip: ipv4 || undefined,
    ipv6: ipv6 || undefined,
    verbose: true,
  };
  const timeoutMs = getTimeoutMs();

  try {
    const response = await http.fetch(DUCKDNS_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "text/plain",
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(timeoutMs),
    });

    const text = await parseTextResponse(response);

    if (!response.ok) {
      return {
        success: false,
        message: duckdnsT("updateFailedWithStatus", {
          status: response.status,
          detail: text || duckdnsT("requestFailed"),
        }),
      };
    }

    const lines = text.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
    const status = lines[0] || text;

    if (status !== "OK") {
      return {
        success: false,
        message: duckdnsT("updateFailed", {
          detail: text || duckdnsT("nonOkResponse"),
        }),
      };
    }

    const result = lines[lines.length - 1];
    const detail = result && result !== "OK" ? ` (${result})` : "";

    return {
      success: true,
      message: duckdnsT("success", { detail }),
      ipv4Updated: !!ipv4,
      ipv6Updated: !!ipv6,
    };
  } catch (error) {
    const err = error instanceof Error ? error : new Error(String(error));
    throw new Error(duckdnsT("requestError", { detail: err.message }));
  }
}

File: apps/server-admin-rs/src/ddns/routes/providers/duckdns.rs
use super::*;

pub(in crate::ddns::routes) fn duckdns_catalog_entry() -> Value {
    provider(
        "duckdns",
        "DuckDNS",
        vec![
            field("domains", "Domains", "text", "home,lab", true),
            field("token", "Token", "password", "DuckDNS Token", true),
        ],
    )
}

pub(in crate::ddns::routes) async fn update_duckdns(
    translator: &Translator,
    config: &HashMap<String, String>,
    http_options: &DDNSHttpClientOptions,
    ipv4: Option<&str>,
    ipv6: Option<&str>,
) -> anyhow::Result<DDNSProviderUpdateResult> {
    let domains = config_value(config, "domains");
    let token = config_value(config, "token");
    if domains.is_empty() || token.is_empty() {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.duckdns.configIncomplete",
            &[],
        )));
    }
    if ipv4.is_none() && ipv6.is_none() {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.duckdns.noIpAvailable",
            &[],
        )));
    }
    let client = ddns_http_client(translator, http_options)
        .map_err(|error| provider_request_error(translator, "duckdns", error))?;
    let response = client
        .post("https://ddns.duckdns.fnknock.cn/")
        .header(reqwest::header::ACCEPT, "text/plain")
        .json(&json!({
            "domains": domains,
            "token": token,
            "ip": ipv4,
            "ipv6": ipv6,
            "verbose": true,
        }))
        .send()
        .await
        .map_err(|error| provider_request_error(translator, "duckdns", error))?;
    let status = response.status();
    let text = response_text(response)
        .await
        .map_err(|error| provider_request_error(translator, "duckdns", error))?;
    if !status.is_success() {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.duckdns.updateFailedWithStatus",
            &[
                ("status", status.as_u16().to_string()),
                (
                    "detail",
                    if text.is_empty() {
                        ddns_text(translator, "providers.duckdns.requestFailed", &[])
                    } else {
                        text
                    },
                ),
            ],
        )));
    }
    let lines = text
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .collect::<Vec<_>>();
    if lines.first().copied() != Some("OK") {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.duckdns.updateFailed",
            &[(
                "detail",
                if text.is_empty() {
                    ddns_text(translator, "providers.duckdns.nonOkResponse", &[])
                } else {
                    text
                },
            )],
        )));
    }
    let detail = lines
        .last()
        .copied()
        .filter(|value| *value != "OK")
        .unwrap_or("");
    Ok(DDNSProviderUpdateResult {
        success: true,
        message: if detail.is_empty() {
            ddns_text(
                translator,
                "providers.duckdns.success",
                &[("detail", String::new())],
            )
        } else {
            ddns_text(
                translator,
                "providers.duckdns.success",
                &[("detail", format!(" ({detail})"))],
            )
        },
    })
}
